cmd/duffle/relocate.go

```diff
@@ -0,0 +1,155 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	"github.com/deislabs/duffle/pkg/loader"
+)
+
+const relocateDesc = `
+Relocates the images referenced by a bundle and creates a new bundle with an updated image map.
+
+The --repository-prefix flag determines the repositories for the relocated images.
+Each image is tagged with a name starting with the given prefix and pushed to the repository.
+
+For example, if the repository-prefix is example.com/user, the image istio/proxyv2 is relocated
+to a name starting with example.com/user/ and pushed to a repository hosted by example.com.
+`
+
+type relocateCmd struct {
+	bundle     string
+	dest       string
+	repoPrefix string
+	out        io.Writer
+	//verbose    bool
+	//insecure   bool
+}
+
+func newRelocateCmd(w io.Writer) *cobra.Command {
+	relocate := &relocateCmd{out: w}
+
+	cmd := &cobra.Command{
+		Use:     "relocate [INPUT-BUNDLE] [OUTPUT-BUNDLE]",
+		Short:   "relocate images in a CNAB bundle",
+		Long:    relocateDesc,
+		Example: `duffle relocate helloworld hellorelocated --repository-prefix example.com/user`,
+		Args:    cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			// validate --repository-prefix if it is set, otherwise allow flag omission to be diagnosed as such
+			if cmd.Flags().Changed("repository-prefix") {
+				if err := FlagValidRepository("repository-prefix")(cmd); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			relocate.bundle = args[0]
+			relocate.dest = args[1]
+
+			return relocate.run()
+		},
+	}
+
+	f := cmd.Flags()
+	f.StringVarP(&relocate.repoPrefix, "repository-prefix", "r", "", "a prefix for relocated image names")
+	cmd.MarkFlagRequired("repository-prefix")
+	//f.BoolVarP(&relocate.verbose, "verbose", "v", false, "Verbose output")
+	//f.BoolVarP(&relocate.insecure, "insecure", "k", false, "Do not verify the bundle (INSECURE)")
+
+	return cmd
+}
+
+func (ex *relocateCmd) run() error {
+	bundlefile, l, err := ex.setup()
+	if err != nil {
+		return err
+	}
+	if err := ex.Relocate(bundlefile, l); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ex *relocateCmd) Relocate(bundlefile string, l loader.Loader) error {
+	return nil
+}
+
+func (ex *relocateCmd) setup() (string, loader.Loader, error) {
+	//bundlefile, err := resolveBundleFilePath(ex.bundle, ex.home.String(), ex.bundleIsFile, ex.insecure)
+	//if err != nil {
+	//	return "", nil, err
+	//}
+	//
+	//l, err := getLoader(ex.home.String(), ex.insecure)
+	//if err != nil {
+	//	return "", nil, err
+	//}
+
+	return "", nil, nil
+}
+
+type FlagsValidator func(cmd *cobra.Command) error
+
+func FlagValidRepository(flagName string) FlagsValidator {
```

Comment:
nit: is this function used outside of the command? Can it be un-exported?

Reply:
No. Yes. Fixed. :-)

Also made `FlagsValidator` private.